### DISCUSSION TITLE
Replacing nginx with YARP

### DISCUSSION THREAD

**Discussion by Alik2015** [dotnet/yarp#1183] [Q&A] [answered]
Would YARP Replace NGINX/Apache?  If so, in our current setup all connections come to port 80 and then nginx manages them.  For example /api/xxx forwards to localhost:1000, /api2/ to localhost:2000 etc.

By default domain.com lands into static html.

Question: Would we run YARP on port 80 by default?  I can't see any tutorials about production deployment ubuntu/linux.  Could someone point to a tutorial showing production setup not dev setup?

**Comment by karelz** [comment]
Triage: @Alik2015 are you looking for docs on how to migrate NGINX config to YARP?
Or are you missing some specific "How to" docs in YARP that you could use?

**Reply by Alik2015** [reply]
Yes looking for docs and how to set up in production and replace nginx.

**Reply by samsp-msft** [reply]
Yes, you can run YARP on port 80. Using the basic sample, in config change [this line](https://github.com/microsoft/reverse-proxy/blob/762b1a43216cb2161a0dcb422036bb6fb9e84351/samples/BasicYarpSample/appsettings.json#L4) to the IP/port combination(s) you want it to listen to.

The [config sample](https://github.com/microsoft/reverse-proxy/blob/main/samples/ReverseProxy.Config.Sample/appsettings.json) shows all the properties that can be set through config, which is similar in concept to most of what you find in NGINX.

**Comment by Alik2015** [comment]
In our setup the default for example is served from var/www and other folders it's static react or Vue html.  How would this be set using routes? On 18 Aug 2021 08:40, Sam Spencer ***@***.***> wrote:
Yes, you can run YARP on port 80. Using the basic sample, in config change this line to the IP/port combination(s) you want it to listen to.
The config sample shows all the properties that can be set through config, which is similar in concept to most of what you find in NGINX.

**Reply by samsp-msft** [reply]
The reverse proxy doesn't care not need to understand where the content is coming from for the destination server - it works purely on the url-space. The following does what you asked for in your original post /api1/* -> localhost:1000, and /api2/* -> localhost:2000.


```json
{
  // Base URLs the server listens on, must be configured independently of the routes below.
  // Can also be configured via Kestrel/Endpoints, see https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints
  "Urls": "http://localhost:80;https://localhost:443",

  //Sets the Logging level for ASP.NET
  "Logging": {
    "LogLevel": {
      "Default": "Information",
      // Uncomment to hide diagnostic messages from runtime and proxy
      // "Microsoft": "Warning",
      // "Yarp" : "Warning",
      "Microsoft.Hosting.Lifetime": "Information"
    }
  },

  "ReverseProxy": {
    // Routes tell the proxy which requests to forward
    "Routes": {
      "api1": {
        "ClusterId": "cluster1",
        "Match": {
          "Path": "/api1/{**catch-all}"
        }
      },
      "route2": {
        // matches /something/* and routes to 2 external addresses
        "ClusterId": "cluster2",
        "Match": {
          "Path": "/api2/{**catch-all}"
        }
      }
    },
    // Clusters tell the proxy where and how to forward requests
    "Clusters": {
      "cluster1": {
        "Destinations": {
          "dest1": {
            "Address": "http://localhost:1000"
          }
        }
      },
      "cluster2": {
        "Destinations": {
          "dest2": {
            "Address": "http://localhost:2000"
          }
        },
      }
    }
  }
}
```

**Reply by Alik2015** [reply]
I appreciate your response.  I believe there might be confusion in me asking the question so will try and clarify.  I understand that I can make this setup and you pointed out I can listen on port 80 which is fine. 

What I was hoping is how the following scenario would be handled.

I match a route lets say /portal or /admin and my apps are not running on localhost:xxx as they are just static html files located somewhere on the server.

Your example explains setup when all things are running on some port and use localhost:xxx.

In our case the apps are not .net apps etc they are just html files that need to be served from a folder and thue API's will be set up as apps like in your example just frontend is static.

So destinations accepts localhost:xxx or url but can it accept a local path?  e.g /var/www/sites/ which it would server index.html?

**Reply by Tratcher** [reply]
While serving static assets is not part of the YARP route configuration, you can do it in code using the ASP.NET Core StaticFiles middleware. Something like: `app.UseStaticFiles("/var/www/sites");`

**Comment by karelz** [comment, marked as answer]
Static files delivery is not mainline scenario yet -- at least not via config. The work is tracked in #187.
